### PR TITLE
refactor `indexing.py`: introduce `.oindex` for Explicitly Indexed Arrays

### PR DESCRIPTION
**Pull request by andersy005**
this pull request represents an initial effort to refactor the indexing logic within `indexing.py` by separating it from `variable.py`. to enhance the flexibility of explicitly indexed array classes, this PR introduces orthogonal ~~and vectorized~~ indexing methods—`.oindex[key]` ~~and `.vindex[key[`~~. these changes are a part of a broader initiative aimed at enabling indexing support for namedarray

- [x] Towards https://github.com/pydata/xarray/issues/8238
- [ ] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

edit 1:
- to minimize the blast radius of this PR, i decided to just focus on `oindex[key]` method in this PR. i intend to introduce `vindex[key]` in a separate PR

## TODO:
- [x] https://github.com/pydata/xarray/pull/8750#discussion_r1491169460
- [ ] Enable passing the underlying tuple key

### CODE REVIEW

File: xarray/core/indexing.py [hunk 1]
```diff
@@ -1417,41 +1430,54 @@ def __init__(self, array):
         """
         self.array = array
 
+    def _rewrite_indexer(self, key):
```
Comment (dcherian) [inline xarray/core/indexing.py:1433]: Let's delete this and close https://github.com/pydata/xarray/issues/8377. There will be a failing test around `.sort` that can also just be deleted.

File: xarray/core/variable.py [hunk 2]
```diff
@@ -761,7 +757,14 @@ def __getitem__(self, key) -> Self:
         array `x.values` directly.
         """
         dims, indexer, new_order = self._broadcast_indexes(key)
-        data = as_indexable(self._data)[indexer]
+        indexable = as_indexable(self._data)
```
Comment (dcherian) [inline xarray/core/variable.py:760]: @shoyer we discussed this a while ago but would be good to get your perspective here again.

We are thinking of preserving the `*Indexer` objects in the returned values from `self._broadcast_indexes`, and then dispatching to `__getitem__`, `oindex` and `vindex` as appropriate on the underlying arrays. We will update all indexing adapters to provide all three methods.

File: xarray/core/variable.py [hunk 3]
```diff
@@ -761,7 +757,14 @@ def __getitem__(self, key) -> Self:
...
+
+        if isinstance(indexer, BasicIndexer):
+            data = indexable[indexer]
+        elif isinstance(indexer, OuterIndexer):
+            data = indexable.oindex(indexer)
+        else:
+            data = indexable.vindex(indexer)
```
Comment (dcherian) [inline xarray/core/variable.py:767]: We should be passing the underlying tuple down to the indexing wrappers
```suggestion
        if isinstance(indexer, BasicIndexer):
            data = indexable[indexer.key]
        elif isinstance(indexer, OuterIndexer):
            data = indexable.oindex(indexer.key)
        else:
            data = indexable.vindex(indexer.key)
```

File: xarray/core/indexing.py [hunk 4]
```diff
@@ -1332,6 +1332,14 @@ def _indexing_array_and_key(self, key):
     def transpose(self, order):
         return self.array.transpose(order)
 
+    def oindex(self, key):
```
Comment (dcherian) [inline xarray/core/indexing.py:1335]: As discussed, it might be easier to do one of these first (either `oindex` or `vindex`) but no strong preference on my end

File: xarray/core/variable.py [hunk 5]
```diff
@@ -761,7 +757,14 @@ def __getitem__(self, key) -> Self:
...
         dims, indexer, new_order = self._broadcast_indexes(key)
-        data = as_indexable(self._data)[indexer]
+        indexable = as_indexable(self._data)
+
+        if isinstance(indexer, BasicIndexer):
+            data = indexable[indexer]
+        elif isinstance(indexer, OuterIndexer):
+            data = indexable.oindex(indexer)
+        else:
+            data = indexable[indexer]
```
Comment (dcherian) [inline xarray/core/variable.py:767]: These should be indexable properties btw:
https://numpy.org/neps/nep-0021-advanced-indexing.html#high-level-changes
```suggestion
        if isinstance(indexer, BasicIndexer):
            data = indexable[indexer]
        elif isinstance(indexer, OuterIndexer):
            data = indexable.oindex[indexer]
        else:
            data = indexable[indexer]
```

File: xarray/core/indexing.py (hunk 1 above)
Comment (andersy005) [reply xarray/core/indexing.py:1435]: Done: 

- https://github.com/pydata/xarray/pull/8758

File: xarray/core/variable.py (hunk 5 above)
Comment (andersy005) [reply xarray/core/variable.py]: Done!

File: xarray/core/variable.py (hunk 3 above)
Comment (andersy005) [reply xarray/core/variable.py]: @dcherian, i looked into this, and it turned out that passing the underlying tuple at this stage (without support for `.vindex`) requires a series of special cases for orthogonal indexing. this results in extensive code duplication between the basic indexer and the vectorized indexer. it might be easier to address this in a separate PR once `vindex` is in place

File: xarray/core/indexing.py [hunk 6]
```diff
@@ -325,6 +325,18 @@ def as_integer_slice(value):
...
+    __slots__ = ("func",)
+
+    def __init__(self, func):
+        self.func = func
+
+    def __getitem__(self, key):
+        return self.func(key)
+
+
 class BasicIndexer(ExplicitIndexer):
```
Comment (dcherian) [inline xarray/core/indexing.py:340]: drive-by comment: Our TODO list should include migrating these indexer objects to `namedarray/indexer.py` or something (eventually)

**Comment by dcherian** [comment]
Do you know why there are no changes to the backend arrays yet? Those too use `*Indexer` objects, and we want to break that out. However now I really there's a backwards compatibility problem there. Can you take a look at cfgrib/rioxarray/tiledb-python to see how they use these lazy arrays and indexing please?

**Comment by andersy005** [comment]
>  Can you take a look at cfgrib/rioxarray/tiledb-python to see how they use these lazy arrays and indexing please?

after talking a look at the respective repos, it's my understanding both the rioxarray and cfrib projects use the `explicit_indexing_adapter` function. this function acts a bridge, connecting high-level *Indexer objects with backend-specific indexing methods. thanks to each backend wrapping its `BackendArray` within a `LazilyIndexedArray` (which supports the `oindex[]`), everything functions smoothly since we are still passing `*Indexer` objects within the `xr.Variable.__getitem__`. however, i anticipate that modifications will be necessary when we begin to directly pass the underlying tuple to methods like `__getitem__`, `oindex`, and `vindex` within `xr.Variable.__getitem__`

on the other hand, TileDB-CF-Py seems to bypass this pattern, handling the __getitem__ method differently by employing its unique logic for processing the underlying tuple. It still wraps its `BackendArray` within a `LazilyIndexedArray` though. 

- https://github.com/TileDB-Inc/TileDB-CF-Py/blob/2c041de8e5f6ae585518a7e5276a6f0a5daee3bc/tiledb/cf/xarray_engine/_array_wrapper.py#L153

- https://github.com/TileDB-Inc/TileDB-CF-Py/blob/2c041de8e5f6ae585518a7e5276a6f0a5daee3bc/tiledb/cf/xarray_engine/_backend_store.py#L83-L99

- https://github.com/corteva/rioxarray/blob/e33616e6b35076749605d5c8d0ff7908e257c2a1/rioxarray/_io.py#L452C1-L456C1
- https://github.com/ecmwf/cfgrib/blob/f95912c838a42fbfc26312927a28c8305258ddb9/cfgrib/xarray_plugin.py#L153-L159






cc @dcherian

File: xarray/core/variable.py (hunk 2 above)
Comment (andersy005) [reply xarray/core/variable.py:760]: @shoyer, the gentlest of bumps on this. have a few minutes to take a look at this? :smile:

File: xarray/core/indexing.py [hunk 7]
```diff
@@ -325,6 +325,18 @@ def as_integer_slice(value):
...
+
+    __slots__ = ("func",)
+
+    def __init__(self, func):
+        self.func = func
+
+    def __getitem__(self, key):
+        return self.func(key)
+
+
```
Comment (dcherian) [inline xarray/core/indexing.py:339]:
```suggestion
        def __setitem__(self, key):
            raise NotImplementedError
```

This will be needed eventually.

File: xarray/core/indexing.py [hunk 8]
```diff
@@ -470,6 +482,13 @@ def __array__(self, dtype: np.typing.DTypeLike = None) -> np.ndarray:
         # Note this is the base class for all lazy indexing classes
         return np.asarray(self.get_duck_array(), dtype=dtype)
 
+    def _oindex(self, key):
```
Comment (dcherian) [inline xarray/core/indexing.py:485]: Should be `_oindex_get` or similar.

**Review by dcherian — COMMENTED** [review]
Thanks, this generally seems like a step in the right direction
